### PR TITLE
Global variable wrappers

### PR DESCRIPTION
**Pull request by Cyberboss**
So, we're never getting rid of globals entirely, they are too performant. What we can do though is wrap them up in a little package that makes them easier to manage. See __DATASTRUCTURES/globals.dm for the nitty gritty.

Old syntax for declaring globals:
```
var/X
var/list/Y
var/datum/genitalia/Z
var/A = 42
var/list/B = list(burn = "witch")
var/datum/genitalia/C = MakeAPenis()
var/hub_password
```

New syntax:
```
GLOBAL_VAR(X)
GLOBAL_LIST(Y)
GLOBAL_DATUM(Z, /datum/genitalia)
GLOBAL_VAR_INIT(A, 42)
GLOBAL_LIST_INIT(B, list(burn = "witch"))
GLOBAL_DATUM_INIT(C, /datum/genitalia, MakeAPenis())
GLOBAL_VAR(hub_password)
GLOBAL_PROTECT(hub_password)
```

All stored under the `GLOB` global

All controllers remain true globals

~~They are still the same old global variables under the hood but they let us use these 4 new procs on /world. (Currently used nowhere due to security implications, except the last one in master/New)~~

~~- `/world/proc/ReadGlobal(global_name)` - Returns the content of global `global_name` or GLOBAL_FIND_FAIL if it doesn't exist~~
~~- `/world/proc/WriteGlobal(global_name, value)` - Writes `value` into the global `global_name`. Returns TRUE on success, FALSE if the global does not exist~~
~~- `/world/proc/ListGlobals()` - Returns an assoc list of all global variables -> values~~
~~- `/world/proc/InitGlobals()` - Sets all globals to their initial values. Operates in declarative order, like we have it currently.~~

This will be ready to be merged when it compiles. I also banned old style globals from travis.

### CODE REVIEW

**Comment by Iamgoofball** [comment]
i thought i was looking at the citadel code for a few moments

**Comment by coiax** [comment]
This is to put all globals under the VV editor, right? And also globals are bad mkay, I assume

**Comment by Cyberboss** [comment]
Logically global variables are bad. But but because BYOND they are also faster than datum vars.

Putting them under the VV editor is now a possiblity, but I haven't done that because there is host sensitive data in some of the globals (comms_key, sql shit)

**Comment by Iamgoofball** [comment]
@Cyberboss just add a hider for the VV menu for it

**Comment by Cyberboss** [comment]
I'll do it in a different PR, I just want to get the shit ported here

**Comment by Iamgoofball** [comment]
sik

**Comment by MrStonedOne** [comment]
cyberboss, just do a global datum and make them all static.

Internally static vars are treated the same as global vars, only they show up on vars[]!

**Comment by Cyberboss** [comment]
Good idea

**Comment by MrStonedOne** [comment]
also, explain to how the hell any of this works.

**Comment by MrStonedOne** [comment]
oh wait..... I see how it works. I don't think this will be as preformant as you think. good thing i had my idea =P

**Comment by Cyberboss** [comment]
Extreme abuse of how you can override the same proc at the same level. I don't expect it to be performant, but at least it's there now

**Comment by MrStonedOne** [comment]
It won't be needed under a singleton `globals` datum setup

**Comment by RemieRichards** [comment]
for the love of god do not call it ```globals```, too much of a clash with the build in ```global``` """object"""

**Comment by Cyberboss** [comment]
Wouldn't even be necessary if it was a real object *grumble,grumble*

**Comment by Cyberboss** [comment]
Anything against just `G`?

**Comment by coiax** [comment]
G might be used in loops for a variable name, I guess. Why not `GLOBAL`? or will that make it look too much like a define?

**Comment by Cyberboss** [comment]
Frankly, prefixing e v e r y global access with anything is a killer task, that's why I designed this system as orthogonal to what we currently have

**Comment by optimumtact** [comment]
I vote for 
SLOTH

because programmers who rely on globals are lazy and I like sloths

**Comment by optimumtact** [comment]
Anyway I'm not sure I see the point of this, globals are annoying but sometimes necessary and I'm not sure what kind of overhead this is going to introduce.

**Comment by Cyberboss** [comment]
@optimumtact In either solution the access overhead is still the same

Both solutions allow all the good shit (VV, programmatic edits, return to initial value, etc..)

The one up now does not require prefixing every global access but is slower when you want to iterate every variable

MSO's solution requires prefixing every access with some new global datum, but is much faster to iterate over.

**Comment by Cyberboss** [comment]
Fuck, my way won't work, it's overflowing the call stack ree

**Comment by Cyberboss** [comment]
I'll go with SLOTH for now, it's unused so I can just grep xargs sed it later

**Comment by Cyberboss** [comment]
God that was painful

**Comment by Cyberboss** [comment]
untitled

**Comment by RemieRichards** [comment]
it should be ```GLOB```
also there's no wrapper anymore, might want to change your pr title.

File: code/__DEFINES/MC.dm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#define MC_TICK_CHECK ( ( world.tick_usage > CURRENT_TICKLIMIT || src.state != SS_RUNNING ) ? pause() : 0 )
+#define MC_TICK_CHECK ( ( world.tick_usage > GLOB.CURRENT_TICKLIMIT || src.state != SS_RUNNING ) ? pause() : 0 )
```
Comment (optimumtact) [inline code/__DEFINES/MC.dm]: glob

**Comment by Cyberboss** [comment]
This is SUCH A GRIND!!!

File: code/__DEFINES/MC.dm (hunk 1 above)
Comment (Cyberboss) [inline code/__DEFINES/MC.dm]: Where?

**Comment by Iamgoofball** [comment]
There should be a way to find these with regex

**Comment by Cyberboss** [comment]
Fucking finished

**Comment by Cyberboss** [comment]
What protected shit looks like 
untitled

**Comment by Cyberboss** [comment]
Broke something 1 sec

**Comment by Cyberboss** [comment]
K done fur realz

File: code/__DEFINES/MC.dm (hunk 1 above)
Comment (F-OS) [inline code/__DEFINES/MC.dm]: OH GOD ORDER LASERS AND EMITTERS

**Comment by Cyberboss** [comment]
Merbeg because confrictur

**Comment by duncathan** [comment]
what could possibly go wrong

**Comment by RemieRichards** [comment]
so this is ok but getters/setters/readwrite traps aren't? rude desu.
@duncathan you're ok with this and since I only need one maintainer to sponsor the PR (oranges was being a butt about it) would you support it pretty please~

**Comment by optimumtact** [comment]
I dont' like this either

**Comment by Cyberboss** [comment]
#24144 <3 @duncathan

**Comment by Core0verload** [comment]
On Animus, globals controller breaks on initialization somehow, breaking the entire round. The bug appears randomly, more often if amount of users is high. Looks like a BYOND bug if I ever seen one.

Why the fuck did you do this? Why switch something as important as global variables to untested BYOND 511 feature? WHYYYYYY?

**Comment by optimumtact** [comment]
Because it works fine for us and your issue has not presented for us, therefore it must be an issue with your setup.

**Comment by optimumtact** [comment]
Still think the boilerplate on this wasn't worth it though.

**Comment by MrStonedOne** [comment]
>Why switch something as important as global variables to untested BYOND 511 feature? WHYYYYYY?

This uses features that have been in since 500

**Comment by Core0verload** [comment]
>This uses features that have been in since 500

Bullshit.

>Variadic #define macros are now supported. For example:
#define dostuff(a,b...) a.myproc(b)
Preceding the variadic argument with ## in the replacement expression will trim off any preceding commas if the argument is empty.

from 511 release notes

**Comment by Core0verload** [comment]
The only thing different in Animus setup is that the server runs Linux. The issue never appears on Windows systems.

**Comment by optimumtact** [comment]
@MrStonedOne redirect sybil to the linux test server and see how long it lasts

**Comment by MrStonedOne** [comment]
it does not use ## because travis still compiles for 510 so travis would fail to compile if that were the case

**Comment by optimumtact** [comment]
It's probably valid syntaxa in 510, if meaningless, which would be pretty funny

**Comment by RemieRichards** [comment]
No, it does do something, ```##``` inserts the variable without a space, so you can do
```/existing/type/path/##some_dynamic_end_path```, the 511 change just makes it trim off commas so you can use it to make lists dynamically too.

**Comment by Cyberboss** [comment]
This pr didn't even need the `##`s, they were purt there for consistency. You can remove them if you want.

**Comment by RemieRichards** [comment]
There's no reason TO remove them either. they work fine.

File: code/modules/events/spacevine.dm [hunk 2]
```diff
@@ -434,14 +434,14 @@
 	var/list/growth_queue = list()
 	var/spread_multiplier = 5
 	var/spread_cap = 30
-	var/list/mutations_list = list()
+	var/list/GLOB.mutations_list = list()
```
Comment (octareenroon91) [inline code/modules/events/spacevine.dm]: #26175

Comment (Cyberboss) [inline code/modules/events/spacevine.dm]: FUCK

**Comment by PJB3005** [comment]
What the unholy eye burning fuck is this?

**Comment by duncathan** [comment]
a namespace for globals

**Comment by PJB3005** [comment]
But why.

**Comment by Cyberboss** [comment]
VV

**Comment by PJB3005** [comment]
> he doesn't create a python script to preprocess the code.

**Comment by MrStonedOne** [comment]
@PJB3005 static vars on objects have the access speed of globals while also being vv-able.

This is a framework that abuses preprocessor macros to do that.

**Comment by PJB3005** [comment]
It's also eye burningly ugly.

**Comment by MrStonedOne** [comment]
form vs function my man.

**Comment by PsiOmegaDelta** [comment]
Hm. Bay has already been able to VV global variables for some time but it seems worth the hassle to port this anyway just for the init ordering (and if I understand things correctly the global var list is automagically populated instead of requiring running an script which is also a plus).

**Pull request merged**